### PR TITLE
remove region_blend_map

### PR DESCRIPTION
**Pull request by Xtarsia**
swap to region blend function, removed region blend map.
removed 2 uniforms in lieu of a fixed blend distance.
~~removed offset from get_region_uv2() as it was causeing an artifact.~~

### CODE REVIEW

**Comment by Xtarsia** [comment]
> > removed offset from get_region_uv2() as it was causeing an artifact.
> 
> Thank you for the update. Is this the artifact making the terrain stepped in the mobile renderer? #314

sadly not.

That issue can be fixed if the new normal method is employed, and the heightmap filtering can be set to nearest, along with removing the 0.5 offset when UV2 is calculated.

**Comment by TokisanGames** [comment]
> -ivec2 pos = ivec2(floor(uv - vec2(_region_texel_size * 0.5))) + (_region_map_size / 2);
> +ivec2 pos = ivec2(floor(uv)) + (_region_map_size / 2);

I see the artifact when this is reverted. This was introduced by PR #348 which I merged based on Fish's recommendation and your approval, but still don't understand the original problem he posed, nor did I see any difference. Please review Fish's arguments there again (related to this offset) and confirm that what he wrote is not relevant or necessary.

**Comment by Xtarsia** [comment]
hmmm actually I think it is wrong. the top image is actually correct, because the bilinear interpolation is (correctly) not occuring across region boundarys.

**Comment by Xtarsia** [comment]
yeah, without it is correct, can can clearly see how the map pixels are centered over the vertices.

image


the color map not blending across regions is technically correct too (and could be fixed via conditional lookups, but not sure its worth the cost, perhaps worth opening an issue for it)

**Comment by TokisanGames** [comment]
From a user perspective, when painting, regions should be invisible.

**Blending**
When painting directly across a region boundary in one swipe, it won't be noticeable.
image

Unknowingly painting along the region boundary will produce an artifact and maybe confusion.
image

The offset didn't fix blending over the region boundary before, so this has always been an issue. But I've never noticed it and it is easily fixable by the user. So I think we can ignore this artifact for now.

**Offset**
However, this does not look correct. 
image

With regions on both sides, the color should go right up to the region boundary so it is consistent. The code may be mathematically correct, but from a user experience, this looks wrong. Looking at the four edges of a region, it's consistently offset by (-0.5, -0.5) vertices on all region boundaries, which the offset does fix. I don't see this stretching.

**Comment by Xtarsia** [comment]
its certainly a confusing issue,

vertex indices go from 0 to 1023 along x/z, so;
The bottom left edge of any region has control/height/color map values from that region.
The top right edge of any region has control/height/color map values from the next region.

imo, for true correctness, the vertex grid region lines (the thick white ones only) should be offset half a texel.

But what looks good to user perspective, does matter more (and smoke and mirrors is what game dev is all about tbh) so i've undone the undo of the undo (?) again and will leave it at that before this gets stuck in the weeds.

File: src/shaders/main.glsl [hunk 1]
```diff
@@ -86,26 +86,22 @@ varying float v_region_border_mask;
 // Takes in UV world space coordinates, returns ivec3 with:
 // XY: (0 to _region_size) coordinates within a region
 // Z: layer index used for texturearrays, -1 if not in a region
-ivec3 get_region_uv(vec2 uv) {
-	uv *= _region_texel_size;
-	ivec2 pos = ivec2(floor(uv)) + (_region_map_size / 2);
-	int bounds = int(pos.x >= 0 && pos.x < _region_map_size && pos.y >= 0 && pos.y < _region_map_size);
+ivec3 get_region_uv(const vec2 uv) {
+	ivec2 pos = ivec2(floor(uv * _region_texel_size)) + (_region_map_size / 2);
+	int bounds = int(uint(pos.x | pos.y) < uint(_region_map_size));
```
Comment (TokisanGames) [inline src/shaders/main.glsl:91]: Wow. I had to think about this one, and I still don't understand how it works if x or y are negative. Did you come up with this?

**Comment by TokisanGames** [comment]
Ah, perhaps we need to rethink where that vertex border should be considering image map pixels, collision, and blending between vertices. I want to think more about it. For now, thanks for removing it so we can merge this one in.

Please remove the `get_region_blend_map()` block from CodeGenerated.gd. And you can also remove the TextureRect from UI.tscn.

I think some people would like to blend in the noise and not have a hard coded moat around their regions. Something along these lines, but also blending within the region a bit. This just blends outside the region.
```
uniform float noise_blend1 : hint_range(0, 0.5) = 0.;
float region_blend(vec2 uv2) {
	...
	//vec2 w = smoothstep(vec2(0.), vec2(1.), fract(uv2));
	vec2 w = smoothstep(vec2(noise_blend1*2.), vec2(1.), fract(uv2)+vec2(noise_blend1));
```

**Comment by Xtarsia** [comment]
Rebased and should be good to go.

**Comment by TokisanGames** [comment]
Great work! Thank you.

**Pull request merged**